File: app/src/main/java/com/micewine/emu/controller/XKeyCodes.java
package com.micewine.emu.controller;

import static com.micewine.emu.controller.ControllerUtils.KEYBOARD;

import android.util.Log;

import java.util.ArrayList;
import java.util.List;

public class XKeyCodes {
    private record ScanKeyCode(String name, int scanCode, int keyCode) {}

    private static final ScanKeyCode[] scanKeyCodes = {
            new ScanKeyCode("Up", 103, 19),
            new ScanKeyCode("Down", 108, 20),
            new ScanKeyCode("Left", 105, 21),
            new ScanKeyCode("Right", 106, 22),
            new ScanKeyCode("ESC", 1, 111),
            new ScanKeyCode("Enter", 28, 66),
            new ScanKeyCode("Space", 57, 62),
            new ScanKeyCode("A", 30, 29),
            new ScanKeyCode("B", 48, 30),
            new ScanKeyCode("C", 46, 31),
            new ScanKeyCode("D", 32, 32),
            new ScanKeyCode("E", 18, 33),
            new ScanKeyCode("F", 33, 34),
            new ScanKeyCode("G", 34, 35),
            new ScanKeyCode("H", 35, 36),
            new ScanKeyCode("I", 23, 37),
            new ScanKeyCode("J", 36, 38),
            new ScanKeyCode("K", 37, 39),
            new ScanKeyCode("L", 38, 40),
            new ScanKeyCode("M", 50, 41),
            new ScanKeyCode("N", 49, 42),
            new ScanKeyCode("O", 24, 43),
            new ScanKeyCode("P", 25, 44),
            new ScanKeyCode("Q", 16, 45),
            new ScanKeyCode("R", 19, 46),
            new ScanKeyCode("S", 31, 47),
            new ScanKeyCode("T", 20, 48),
            new ScanKeyCode("U", 22, 49),
            new ScanKeyCode("V", 47, 50),
            new ScanKeyCode("W", 17, 51),
            new ScanKeyCode("X", 45, 52),
            new ScanKeyCode("Y", 21, 53),
            new ScanKeyCode("Z", 44, 54),
            new ScanKeyCode("'", 40, 75),
            new ScanKeyCode("LCtrl", 29, 113),
            new ScanKeyCode("RCtrl", 97, 114),
            new ScanKeyCode("LShift", 42, 59),
            new ScanKeyCode("RShift", 54, 60),
            new ScanKeyCode("Tab", 15, 61),
            new ScanKeyCode("AltLeft", 56, 57),
            new ScanKeyCode("F1", 59, 131),
            new ScanKeyCode("F2", 60, 132),
            new ScanKeyCode("F3", 61, 133),
            new ScanKeyCode("F4", 62, 134),
            new ScanKeyCode("F5", 63, 135),
            new ScanKeyCode("F6", 64, 136),
            new ScanKeyCode("F7", 65, 137),
            new ScanKeyCode("F8", 66, 138),
            new ScanKeyCode("F9", 67, 139),
            new ScanKeyCode("F10", 68, 140),
            new ScanKeyCode("F11", 87, 141),
            new ScanKeyCode("F12", 88, 142),
            new ScanKeyCode("Insert", 110, 124),
            new ScanKeyCode("Home", 102, 122),
            new ScanKeyCode("PageUp", 104, 92),
            new ScanKeyCode("Delete", 111, 112),
            new ScanKeyCode("End", 107, 123),
            new ScanKeyCode("PageDown", 109, 93),
            new ScanKeyCode("BackSpace", 14, 67),
            new ScanKeyCode("0", 11, 7),
            new ScanKeyCode("1", 2, 8),
            new ScanKeyCode("2", 3, 9),
            new ScanKeyCode("3", 4, 10),
            new ScanKeyCode("4", 5, 11),
            new ScanKeyCode("5", 6, 12),
            new ScanKeyCode("6", 7, 13),
            new ScanKeyCode("7", 8, 14),
            new ScanKeyCode("8", 9, 15),
            new ScanKeyCode("9", 10, 16),
    };

    public static List<String> getKeyNames(boolean getMouseButtons) {
        final ArrayList<String> keyNames = new ArrayList<>();

        keyNames.add("--");

        if (getMouseButtons) {
            keyNames.add("M_Left");
            keyNames.add("M_Middle");
            keyNames.add("M_Right");
            keyNames.add("M_WheelUp");
            keyNames.add("M_WheelDown");
        } else {
            keyNames.add("Mouse");
        }

        for (ScanKeyCode scanKeyCode : scanKeyCodes) {
            keyNames.add(scanKeyCode.name);
        }

        return keyNames;
    }

    public static ButtonMapping getMapping(String key) {
        ScanKeyCode mapping = null;

        for (ScanKeyCode scanKeyCode : scanKeyCodes) {
            if (scanKeyCode.name.equals(key)) {
                mapping = scanKeyCode;
            }
        }

        if (mapping == null) {
            return new ButtonMapping(key);
        }

        return new ButtonMapping(key, mapping.scanCode, mapping.keyCode, KEYBOARD);
    }

    public static class ButtonMapping {
        String name;
        int scanCode;
        int keyCode;
        int type;

        public ButtonMapping() {
            this.name = "";
            this.scanCode = -1;
            this.keyCode = -1;
            this.type = -1;
        }

        public ButtonMapping(String name) {
            this.name = name;
            this.scanCode = -1;
            this.keyCode = -1;
            this.type = -1;
        }

        public ButtonMapping(String name, int scanCode, int keyCode, int type) {
            this.name = name;
            this.scanCode = scanCode;
            this.keyCode = keyCode;
            this.type = type;
        }

        public void setName(String name) {
            this.name = name;
        }

        public String getName() {
            return name;
        }
    }
}

File: app/src/main/java/com/micewine/emu/controller/XKeyCodes.kt
package com.micewine.emu.controller

object XKeyCodes {
    private val scanKeyCodes = mapOf(
        "Up" to (103 to 19),
        "Down" to (108 to 20),
        "Left" to (105 to 21),
        "Right" to (106 to 22),
        "ESC" to (1 to 111),
        "Enter" to (28 to 66),
        "Space" to (57 to 62),
        "A" to (30 to 29),
        "B" to (48 to 30),
        "C" to (46 to 31),
        "D" to (32 to 32),
        "E" to (18 to 33),
        "F" to (33 to 34),
        "G" to (34 to 35),
        "H" to (35 to 36),
        "I" to (23 to 37),
        "J" to (36 to 38),
        "K" to (37 to 39),
        "L" to (38 to 40),
        "M" to (50 to 41),
        "N" to (49 to 42),
        "O" to (24 to 43),
        "P" to (25 to 44),
        "Q" to (16 to 45),
        "R" to (19 to 46),
        "S" to (31 to 47),
        "T" to (20 to 48),
        "U" to (22 to 49),
        "V" to (47 to 50),
        "W" to (17 to 51),
        "X" to (45 to 52),
        "Y" to (21 to 53),
        "Z" to (44 to 54),
        "'" to (40 to 75),
        "LCtrl" to (29 to 113),
        "RCtrl" to (97 to 114),
        "LShift" to (42 to 59),
        "RShift" to (54 to 60),
        "Tab" to (15 to 61),
        "AltLeft" to (56 to 57),
        "F1" to (59 to 131),
        "F2" to (60 to 132),
        "F3" to (61 to 133),
        "F4" to (62 to 134),
        "F5" to (63 to 135),
        "F6" to (64 to 136),
        "F7" to (65 to 137),
        "F8" to (66 to 138),
        "F9" to (67 to 139),
        "F10" to (68 to 140),
        "F11" to (87 to 141),
        "F12" to (88 to 142),
        "Insert" to (110 to 124),
        "Home" to (102 to 122),
        "PageUp" to (104 to 92),
        "Delete" to (111 to 112),
        "End" to (107 to 123),
        "PageDown" to (109 to 93),
        "BackSpace" to (14 to 67),
        "0" to (82 to 144),
        "1" to (79 to 145),
        "2" to (80 to 146),
        "3" to (81 to 147),
        "4" to (75 to 148),
        "5" to (76 to 149),
        "6" to (77 to 150),
        "7" to (71 to 151),
        "8" to (72 to 152),
        "9" to (73 to 153),
    )

    fun getKeyNames(): MutableList<String> {
        val keyNames: MutableList<String> = mutableListOf("Null")

        for (i in scanKeyCodes.keys) {
            keyNames.plusAssign(i)
        }

        return keyNames
    }

    fun getXKeyScanCodes(key: String): MutableList<Int> {
        val scanCode = scanKeyCodes[key]?.first ?: 0
        val keyCode = scanKeyCodes[key]?.second ?: 0
        return mutableListOf(scanCode, keyCode, 0)
    }
}
